code/scripting/api/objs/briefing.h

```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+#include "mission/missionbriefcommon.h"
+#include "mission/missiongoals.h"
+#include "scripting/ade_api.h"
+
+namespace scripting {
+namespace api {
+
+DECLARE_ADE_OBJ(l_BriefStage, brief_stage);
+
+DECLARE_ADE_OBJ(l_Brief, briefing);
+
+DECLARE_ADE_OBJ(l_Goals, mission_goal);
```

Comment:
This is an issue.
We really don't want to create lua objects based on complex types.
This creates copies in lua, and will cause multiple issues, such as
1. Being slow
2. Any changes will not be propagated back to FSO
3. The object won't pick up changes made in FSO

To fix this, we need to make some new object types. l_Brief can just use int instead of the briefing class, and store only the briefing index. See for example the team.h/.cpp API classes as to how that's done.
l_BriefStage should IMO take a class that contains a brief_stage pointer. Since neither brief stages nor briefings are in dynamically allocated memory, that should be safe. An example how this needs to look is waypointlist_h in waypoint.h (just without the name business for the briefing stage)
l_Goals can be an int again, like l_Brief.

Reply:
Can you clarify? I duplicated exactly what m!m did for Cmd Briefs. These don't ever change during runtime.

Reply:
Right, and I think that cmd_briefs need to be changed as well.
What happens here, is that every time you get a briefing stage from lua, it will make a _copy_ of the briefing stage. A copy which is unrelated to the one from FSO, afterwards. As long as you are guaranteed that the Lua API will never change anything in that copy, _and_ that FSO wil never modify the briefing stage while you have your copy in lua, that is fine-ish. But if you can't guarantee that, it becomes a problem. This also implies, that you must pay attention that no one ever implements anything which might be able to modify briefings from Lua, which has reasonable use cases IMO.
I will make a PR later today which'll fix this and the documentation strings for cmd_briefs, then you can see what'd need to be changed.

Reply:
See #4646

Reply:
Ok, excellent. That gives me an example to follow. I'll see if I can figure this out.

Reply:
Ok, I think I've set this up correctly and my tests show it still works as expected.